keps/sig-cli/kubectl-apply-prune-beta.md

```diff
@@ -0,0 +1,141 @@
+
+Here is an example of root ConfigMap object. It has a predictable name or a label `kubectl.kubernetes.io/applyRoot: true` indicating that it is a root object.
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+ name: predictable-name-foo
+ labels:
+   kubectl.kubernetes.io/applyRoot: true
+data:
+ object1:  namespace1:group1:version1:kind1:somename
+ object2:  namespace1:group2:version2:kind2:secondname
+ object3:  namespace3:group3:version3:kind3:thirdname
+```
+
+The presence of a root object indicates apply need to handle prune. When kubectl apply is given such a root object, it finds the live root object with the same name. 
+- If a live root object is not found, Kubectl will create the root object and the corresponding set of objects
+- If a live root object is found, then the set of live objects will be found by reading the data field of the live root object. Kubectl compares the live set with new set of objects and takes following steps: 
+  - set the root object data as the union of two sets and update the root object
+  - create new objects
+  - prune old objects
+  - remove the pruned objects from root object’s data field
+
+
+### Child Objects
+The objects passed to apply will have labels or annotations pointing back to the root object for the same apply.
+```
+ labels:
+    kubectl.kubernetes.io/applyOwner: v1:ConfigMap:predictable-name-foo
+```
+The presence of this label is to confirm that a live object is safe to delete when it is to be pruned.
+
+### Risks and Mitigations
+
+Deciding the correct set of objects to change is hard. `--prune` Alpha tended to solve that problem in kubectl itself. In this new approach, kubectl is not going to decide such set. Instead it delegates the job to users or tools. Those who need to use `--prune` will be responsible to create a root object as a record of the correct set of objects. Kubectl consumes this root object and triggers the pruning action.
+
+This approach requires users/tools to do more work than just specifying a label. which might discourage the uage of `--prune`. On the other side, existing tools such as Kustomize, Helm can utilize this approach by generating a root object when expanding or generating the resources.
+
+## Graduation Criteria
+`--prune` can be moved from Beta to GA when
+
+- There is no severe issue in Beta `--prune`.
+ 
+- `kubectl apply --prune` is widely used in CI workflow with Kubernetes.
+
+## Implementation History
+
+The major implementation includes following steps:
+
+- add functions to identify root object
+- add a different code path to handle the case when root object is present
+- add unit test
+- add integration test
```

Comment:
Should we prioritize end-to-end tests rather than integration tests?

Reply:
Yes, you're right. I'll update it.